client-common/src/main/java/com/cloudera/livy/client/common/ClientConf.java

```diff
@@ -193,4 +212,40 @@ private void check(boolean test, String message, Object... args) {
     }
   }
 
+  /** Logs a warning message if the given config key is deprecated. */
+  private void logDeprecationWarning(String key) {
+    DeprecatedConf altConfs = getConfigsWithAlternatives().get(key);
+    if (altConfs != null) {
+      LOG.warn("The configuration key " + altConfs.key() + " has been deprecated as of Livy "
+        + altConfs.version() + " and may be removed in the future. Please use the new key "
+        + key + " instead.");
+      return;
+    }
+
+    DeprecatedConf depConfs = getDeprecatedConfigs().get(key);
+    if (depConfs != null) {
+      LOG.warn("The configuration key " + depConfs.key() + " has been deprecated as of Livy "
+        + depConfs.version() + " and may be removed in the future. "
+        + depConfs.deprecationMessage());
+    }
+  }
+
+  /** Maps valid key to DeprecatedConf with the deprecated key. */
+  public abstract Map<String, DeprecatedConf> getConfigsWithAlternatives();
+
+  /** Maps deprecated key to DeprecatedConf with the same key. */
+  public abstract Map<String, DeprecatedConf> getDeprecatedConfigs();
```

Comment:
I'm not sure it would make it simpler, I think it would just move the complexity. Rather than two lists with different meanings you would instead have one list with two different types of content. I think the code required to combine the lists would be more complex than separate lists. The first indicator of this I see is that the lists are hashed on different values; one is the deprecated key the other is the new key. A second point would be their different usage, the alternates list is used to link old confs to their replacements, whereas the deprecated list is solely for logging warnings if set. If this was all scala code I would agree that combining would be simpler, but I can't see how in Java. I'm still open doing it if you could point out a simple code change to combine them, I admit I may just be too close to see it.